adr/ADR-43.md

```diff
@@ -0,0 +1,85 @@
+# JetStream Per-Message TTL
+
+| Metadata | Value                     |
+|----------|---------------------------|
+| Date     | 2024-07-11                |
+| Author   | @ripienaar                |
+| Status   | Approved                  |
+| Tags     | jetstream, client, server |
+
+## Context and motivation
+
+Streams support a one-size-fits-all approach to message TTL based on the MaxAge setting. This causes any message in the 
+Stream to expire at that age.
+
+There are numerous uses for a per-message version of this limit, some listed below:
+
+ * KV tombstones are a problem in that they forever clog up the buckets with noise, these could have a TTL to make them expire once not useful anymore
+ * Server-applied limits can result in tombstones with a short per message TTL so that consumers can be notified of limits being processed. Useful in KV watch scenarios being notified about TTL removals
+ * A stream may have a general MaxAge but some messages may have infinite retention, think a schema or type hints in a KV bucket that is forever while general keys have TTLs
+
+Related issues [#3268](https://github.com/nats-io/nats-server/issues/3268)
+
+## Per-Message TTL
+
+We will allow a message to supply a TTL using a header called `Nats-TTL` followed by the duration as seconds.
+
+The duration will be used by the server to calculate the deadline for removing the message based on its Stream 
+timestamp and the stated duration.
+
+The TTL may not exceed the Stream MaxAge to give administrators a hard cap for the purposes of managing resources. 
+The shortest allowed TTL would be 1 second.
+
+A TTL of zero will be ignored, any other unparsable value will result in a error reported in the Pub Ack and the message
+being discarded.
+
+Setting the header `Nats-No-Expire` to `1` will result in a message that will never be expired.
```

Comment:
It might be a good idea to have a setting enabling/disabling `Nats-No-Expire` to give administrators control over abusing it.

Reply:
This seems to contradict the above statement:

> The TTL may not exceed the Stream MaxAge to give administrators a hard cap for the purposes of managing resources.

If no `MaxAge` is set on the stream, then _not_ setting the `Nats-No-Expire` would mean is never expires. If a `MaxAge` is set then `Nats-No-Expire` should not be allowed since that violate the hard cap.

Reply:
We specifically want the ability to keep messages forever.  Imagine a KV where you need to recover the data type for messages in the stream but want the data in general to expire.

These configuration style messages must persist forever while the bulk of the data is limits managed.

As @Jarema says we might need to make this uncapped special messages a boolean but I think we can do that later once we get some feedback